Implement helpers for using alloca-based stack arrays

zah opened the pull request:
There is one open question:

```
  # XXX: is it possible to perform a stack size check before calling `alloca`?
  # On thread init, Nim may record the base address and the capacity of the stack,
  # so in theory we can verify that we still have enough room for the allocation.
  # Research this.
```
Some reference info:
https://stackoverflow.com/questions/46743669/why-does-alloca-not-check-if-it-can-allocate-memory

yglukhov commented:
The question is what do we do if the stack size verification fails. Fail? Fallback to seq? The worst thing about this is that the failure is not deterministic in general case (read: hard to debug).

cheatfate commented on ranges/stackarrays.nim:
```diff
@@ -0,0 +1,57 @@
+  arr.bufferLen = size
+  arr
+
+template toOpenArray*(a: StackArray): auto =
```
This procedure converts whole array, but i think it would be better to have template which supports slices too.

cheatfate commented on ranges/stackarrays.nim:
```diff
@@ -0,0 +1,57 @@
+
+type
+  StackArray*[T] = ptr object
+    bufferLen: int
```
I think `int` is overhead here, because its stack array even `int32` would be enough.

cheatfate commented on ranges/stackarrays.nim:
```diff
@@ -0,0 +1,57 @@
+
+template toOpenArray*(a: StackArray): auto =
+  toOpenArray(a.buffer, 0, a.high)
+
```
I think `toPtr` function with starting index would be also very helpful.

zah replied:
obtaining a pointer is supported in the usual way: `addr arr[i]`

zah commented:
All comments have been addressed. I've added a large disclaimer at the top of the `stackarrays` file, because we've identified several significant risks of using the library.

Also added various possible future directions that I discussed with @yglukhov in Slack.
